text/287-cli-deprecation-warnings.md

```diff
@@ -0,0 +1,224 @@
+
+* Setting the environment variable `CDK_DEPRECATED_IGNORE` will silence these warnings
+  (they will no longer be printed to the console).
+
+* Setting the environment variable `CDK_DEPRECATED_ERROR` will instead fail with an exception when any deprecated element is used.
+
+### Contributing guide
+
+To deprecate an element (class, interface, property, function, etc.) in the CDK,
+you need to add the `@deprecated` JSDoc tag to the documentation block of that element.
+This will change the generated JavaScript code to make it emit a warning to the standard error stream whenever that element is invoked.
+
+## FAQ
+
+### What are we launching today?
+
+In the newest release of the AWS CDK Construct Library,
+any usage of a deprecated element  (class, interface, property, function)
+in your CDK code will result in a warning being printed in the console output.
+
+### Does this mean my code will stop working after this change?
+
+No! This feature only adds warnings to the CLI,
+it does not result in any code that was previously working to now fail.
+
+Note that if you want to migrate to `V2` of the CDK,
+which we strongly recommend,
+you will have to handle all the warnings that this feature emits,
+as all deprecated elements in `V1` will be removed from `V2`,
+and thus no longer available to be used by your code.
+
+If you want to make sure your code does not use any deprecated APIs,
+and thus is ready for migrating to CDK `V2`,
+you can set the `CDK_DEPRECATED_ERROR` environment variable,
+which will make any CDK command that invokes synthesis
+(`cdk synth`, `cdk deploy`, `cdk diff`, etc.)
+fail with an exception if any deprecated element is used.
+
+## Internal FAQ
+
+### Why are we doing this?
+
+We are doing this to help customers have a smoother migration from `V1` to `V2` of the CDK.
+
+### Why should we _not_ do this?
+
+I see three reasons why we might not want to implement this feature:
+
+1. Deprecated elements are already highlighted by virtually any editor/IDE.
+  In my experience, customers are diligent in moving away from deprecated APIs,
+  so this feature might not provide much value.
+2. We risk overloading customers with warnings,
+  training them to ignore our output if there's too much of it.
+  A warning must be read in order to be effective, after all.
+3. Depending on the exact solution chosen
+  (see discussion below),
+  we might not be able to distinguish from the user's code using a deprecated API,
+  and from our own libraries using deprecated APIs.
+  Which means a user might get warnings that they will be unable to get rid of
+  (because they're coming from code that they don't control).
+
+### What changes are required to enable this change?
+
+There are two high-level components of this change:
+
+#### 1. Changes to JSII
+
+We will add a feature to JSII that adds the warning code to the emitted JavaScript for deprecated elements.
+It will be off by default, and will have to be activated explicitly to take effect.
```

Comment:
What does that mean?

The default behavior of the application will be `ignore`, and the CLI will in turn flip that to `warn` by default unless ignored?

Motivation: unit tests etc I suppose?

Reply:
It means the code adding the warnings will not be added to the emitted JavaScript by default at all.